### PR TITLE
Replace ArrayType::castToArrayKey() with Type::toArrayKey()

### PR DESCRIPTION
**Pull request by rvanvelzen**

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
All aboard the Type refactoring train! 🚆  I can't wait to get rid of inheritance in Type classes.

File: src/Type/StringType.php [hunk 1]
```diff
@@ -137,6 +137,11 @@ public function toArray(): Type
 		);
 	}
 
+	public function toArrayKey(): Type
+	{
+		return $this;
```
Comment (herndlm) [inline src/Type/StringType.php:142]: This needs to return most likely `int|string` I believe?

Comment (ondrejmirtes) [reply src/Type/StringType.php:142]: It'd be correct but annoying.

Comment (rvanvelzen) [reply src/Type/StringType.php:142]: It didn't do so in the previous implementation, and though technically that is correct I don't think changing it now is wise.

Comment (herndlm) [reply src/Type/StringType.php:142]: ah right, I thought the failing tests were related, but looks like they aren't :)

**Comment by rvanvelzen** [comment]
The improvement actually causes some tests to fail legitimately, which is why I kept is as a draft. I'll improve those separately first so this can land cleanly :)

**Comment by herndlm** [comment]
ah right, I introduced usage of the old method again in the meantime in `HasOffsetValueType`, sorry 😊

File: src/Type/ArrayType.php [hunk 2]
```diff
@@ -432,55 +429,15 @@ public function toArray(): Type
...
 
-	/** @deprecated Use getArraySize() instead */
-	public function count(): Type
+	public function toArrayKey(): Type
 	{
-		return $this->getArraySize();
+		return new ErrorType();
 	}
 
-	public static function castToArrayKeyType(Type $offsetType): Type
```
Comment (ondrejmirtes) [inline src/Type/ArrayType.php:441]: Removing the method is a BC break, we should keep it but simply call `toArrayKey` on the type. And mark it as deprecated.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**